crates/uv/tests/python_pin.rs

```diff
@@ -229,6 +230,186 @@ fn python_pin_no_python() {
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    error: The requested Python version `3.10` is incompatible with the project `Requires-Python` requirement of `>=3.11`.
+    "###);
+
+    // Request a implementation version that is incompatible
+    uv_snapshot!(context.filters(), context.python_pin().arg("cpython@3.10"), @r###"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    error: The requested Python version `cpython@3.10` is incompatible with the project `Requires-Python` requirement of `>=3.11`.
+    "###);
+
+    // Request a complex version range that resolves to an incompatible version
+    uv_snapshot!(context.filters(), context.python_pin().arg(">3.8,<3.11"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    Pinned `.python-version` to `>3.8, <3.11`
+
+    ----- stderr -----
+    warning: The requested Python version `>3.8, <3.11` resolves to `3.10.[X]` which  is incompatible with the project `Requires-Python` requirement of `>=3.11`.
+    "###);
+
+    uv_snapshot!(context.filters(), context.python_pin().arg("3.11"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    Updated `.python-version` from `>3.8, <3.11` -> `3.11`
+
+    ----- stderr -----
+    "###);
+
+    // Request a implementation version that is compatible
+    uv_snapshot!(context.filters(), context.python_pin().arg("cpython@3.11"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    Updated `.python-version` from `3.11` -> `cpython@3.11`
+
+    ----- stderr -----
+    "###);
+
+    let python_version =
+        fs_err::read_to_string(context.temp_dir.join(PYTHON_VERSION_FILENAME)).unwrap();
+    insta::with_settings!({
+        filters => context.filters(),
+    }, {
+        assert_snapshot!(python_version, @r###"
+        cpython@3.11
+        "###);
+    });
+
+    // Updating `requires-python` should affect `uv python pin` compatibilities.
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "project"
+        version = "0.1.0"
+        requires-python = ">=3.12"
+        dependencies = ["iniconfig"]
+        "#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.python_pin(), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    cpython@3.11
+
+    ----- stderr -----
+    warning: The pinned Python version `cpython@3.11` is incompatible with the project `Requires-Python` requirement of `>=3.12`.
+    "###);
+
+    // Request a implementation that resolves to a compatible version
+    uv_snapshot!(context.filters(), context.python_pin().arg("cpython"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    Updated `.python-version` from `cpython@3.11` -> `cpython`
+
+    ----- stderr -----
+    warning: The requested Python version `cpython` resolves to `3.10.[X]` which  is incompatible with the project `Requires-Python` requirement of `>=3.12`.
```

Comment:
Am I misunderstood or `cpython` must be resolved to 3.11? since context already contains it.

Reply:
ref https://github.com/astral-sh/uv/pull/5205

Reply:
Yes that's problematic. cc @charliermarsh looks like this regressed.

Reply:
This actually isn't a regression or related to #5205, we're just taking the first Python on the path. The problem here is that we need to respect `Requires-Python` on top of pinned version requests (😭).